### PR TITLE
Add azure tables application demo for java

### PR DESCRIPTION
**Pull request by v-hongli1**
Add azure tables application demo for java, with reference to repo [msdocs-azure-data-tables-sdk-dotnet](https://github.com/Azure-Samples/msdocs-azure-data-tables-sdk-dotnet)

@mario-guerra, @joshfree, @annatisch, @mayurid and @vcolin7 for notification.

### CODE REVIEW

**Comment by v-xuto** [comment]
@mario-guerra Please help add the corresponding person to review the PR. Thanks.

**Comment by mario-guerra** [comment]
@v-xuto I've asked our Java SDK developer to review this commit, he said he would do that today.

**Comment by mario-guerra** [comment]
@v-xuto  can you grant @vcolin7 permissions to the repo so he can review the commit?

File: 1-starter-app/AzureTablesDemoApplication/src/main/java/com/data/tables/common/Constants.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package com.data.tables.common;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Constants {
```
Comment (vcolin7) [inline 1-starter-app/AzureTablesDemoApplication/src/main/java/com/data/tables/common/Constants.java]: Based on what I see, constants are not really used in more than one class at a time:
- `SUCCESS_CODE` and `SUCCESS_MSG` are used exclusively in `TablesServiceController`
- `FAILED_CODE` is used exclusively in `GlobalExceptionHandler`
- `BLANK_STRING`, `EXCLUDE_TABLE_ENTITY_KEYS` and `EXCLUDE_TABLE_ENTITY_KEYS` are used exclusively in `WeatherDataUtils`

I think we should simply make these constants in the files they're relevant in and remove this file, unless we are planning in them ever being used in multiple places.

File: 1-starter-app/AzureTablesDemoApplication/src/main/java/com/data/tables/controller/TablesServiceController.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+    private static final String LAYOUT_PAGE = "layouts/_layout";
+
+    private static final String PRIVACY_PAGE = "privacy";
+
+    private static final String INDEX_PAGE = "index";
+
+    private static final String FILTER_PAGE = "filter-results";
+
+    private static final String HOME_PAGE_TITLE = "Home page";
```
Comment (vcolin7) [inline 1-starter-app/AzureTablesDemoApplication/src/main/java/com/data/tables/controller/TablesServiceController.java]: nit;
```suggestion
    private static final String HOME_PAGE_TITLE = "Home Page";
```

File: 1-starter-app/AzureTablesDemoApplication/src/main/java/com/data/tables/controller/TablesServiceController.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        List<WeatherDataModel> entitiesList = tablesService.getFilteredRows(model);
+        return new JSONObject(new HashMap<String, Object>(){{
+            put("entitiesList", entitiesList);
+            put("listOfKeys", WeatherDataUtils.getListOfKeys(entitiesList));
+            put("code", Constants.SUCCESS_CODE);
+            put("msg", Constants.SUCCESS_MSG);
+        }}).toString();
+    }
+
+    @PostMapping("/removeEntity")
```
Comment (vcolin7) [inline 1-starter-app/AzureTablesDemoApplication/src/main/java/com/data/tables/controller/TablesServiceController.java]: We should be consistent in naming these CRUD endpoints. Let's pick between `TableEntity` or simply `Entity` in all of them.

File: 1-starter-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/ITablesService.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import com.data.tables.models.WeatherInputModel;
+import java.util.List;
+
+public interface ITablesService {
+
+    List<WeatherDataModel> getAllRows();
+
+    List<WeatherDataModel> getFilteredRows(FilterResultsInputModel model);
+
+    void removeEntity(WeatherInputModel model);
```
Comment (vcolin7) [inline 1-starter-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/ITablesService.java]: We should consistently name these methos either using `TableEntity` or simply `Entity`.

**Comment by v-xuto** [comment]
@mario-guerra I don't know why, I couldn't find him when adding reviewers.
image
But, I have added read permission in Setting `Manage access`. I'm sure if he can review the code by doing this?
image

File: 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    private TableClient tableClient;
+    /**
+     * Get all weather data entities
+     * @return List<WeatherDataModel>
+     */
+    @Override
+    public List<WeatherDataModel> getAllRows() {
+        List<WeatherDataModel> modelList = new ArrayList<>();
+        tableClient.listEntities().stream().forEach(entity -> modelList.add(WeatherDataUtils.mapTableEntityToWeatherDataModel(entity)));
+        return Collections.unmodifiableList(WeatherDataUtils.filledValue(modelList));
```
Comment (vcolin7) [inline 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java]: nit; We can make this look a bit cleaner by doing the following:
```suggestion
        List<WeatherDataModel> modelList = tableClient.listEntities().stream()
            .map(WeatherDataUtils::mapTableEntityToWeatherDataModel)
            .collect(Collectors.toList());
        return Collections.unmodifiableList(WeatherDataUtils.filledValue(modelList));
```

File: 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    /**
+     * Insert a new expandable weather data
+     */
+    @Override
+    public void insertExpandableData(ExpandableWeatherObject model){
+        tableClient.createEntity(WeatherDataUtils.createTableEntity(model));
+    }
+
+    /**
+     * Update the expandable weather data
```
Comment (vcolin7) [inline 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java]:
```suggestion
     * Insert or update an expandable weather data object
```

File: 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    @Override
+    public void updateEntity(UpdateWeatherObject model) {
+        TableEntity tableEntity = tableClient.getEntity(model.getStationName(), model.getObservationDate());
+        Map<String, Object> propertiesMap = model.getPropertyMap();
+        propertiesMap.keySet().forEach(key -> tableEntity.getProperties().put(key, propertiesMap.get(key)));
+        tableClient.updateEntity(tableEntity);
+    }
+
+    /**
+     * Insert a new expandable weather data
```
Comment (vcolin7) [inline 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java]:
```suggestion
     * Insert a new expandable weather data object
```

File: 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+    /**
+     * Insert a new weather data
+     */
+    @Override
+    public void insertTableEntity(WeatherInputModel model) {
+        tableClient.createEntity(WeatherDataUtils.createTableEntity(model));
+    }
+
+    @Override
```
Comment (vcolin7) [inline 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java]: The Javadoc comment for this method is missing.
```suggestion
    /**
     * Insert or update weather data
     */
    @Override
```

File: 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+     * Delete weather data by table key
+     */
+    @Override
+    public void removeEntity(WeatherInputModel model) {
+        tableClient.deleteEntity(model.getStationName(),
+                WeatherDataUtils.formatRowKey(model.getObservationDate(), model.getObservationTime()));
+    }
+
+    /**
+     * Insert a new weather data
```
Comment (vcolin7) [inline 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java]:
```suggestion
     * Insert new weather data
```

File: 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+        tableClient.createEntity(WeatherDataUtils.createTableEntity(model));
+    }
+
+    @Override
+    public void upsertTableEntity(WeatherInputModel model) {
+        tableClient.upsertEntity(WeatherDataUtils.createTableEntity(model));
+    }
+
+    /**
+     * Update the weather data
```
Comment (vcolin7) [inline 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java]:
```suggestion
     * Update weather data
```

File: 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    /**
+     * Update the expandable weather data
+     */
+    @Override
+    public void upsertExpandableData(ExpandableWeatherObject model){
+        tableClient.upsertEntity(WeatherDataUtils.createTableEntity(model));
+    }
+
+    /**
+     * Insert or update weather data
```
Comment (vcolin7) [inline 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java]:
```suggestion
     * Insert or update multiple instances of weather data using a transaction
```

File: 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+            filters.add(String.format("Precipitation ge %f", model.getMinPrecipitation()));
+        }
+        if (model.getMaxPrecipitation() != null) {
+            filters.add(String.format("Precipitation le %f", model.getMaxPrecipitation()));
+        }
+
+        List<WeatherDataModel> modelList = new ArrayList<>();
+        tableClient.listEntities(new ListEntitiesOptions().setFilter(String.join(" and ", filters)), null, null)
+                .stream().forEach(entity -> modelList.add(WeatherDataUtils.mapTableEntityToWeatherDataModel(entity)));
+        return Collections.unmodifiableList(WeatherDataUtils.filledValue(modelList));
```
Comment (vcolin7) [inline 2-completed-app/AzureTablesDemoApplication/src/main/java/com/data/tables/service/impl/TablesServiceImpl.java]: nit; I also think we can make this look a bit cleaner:
```suggestion
        List<WeatherDataModel> modelList = tableClient.listEntities(new ListEntitiesOptions()
                .setFilter(String.join(" and ", filters)), null, null).stream()
            .map(WeatherDataUtils::mapTableEntityToWeatherDataModel)
            .collect(Collectors.toList());
        return Collections.unmodifiableList(WeatherDataUtils.filledValue(modelList));
```

File: README.md [hunk 13]
```diff
@@ -1,57 +1,52 @@
-# Project Name
+# azure-data-tables sample application
 
-(short, 1-3 sentenced, description of the project)
+This is the sample project for using the [azure-data-tables SDK](https://search.maven.org/artifact/com.azure/azure-data-tables) library with the Azure Cosmos DB Table API.  The quickstart for this application is located at [https://docs.microsoft.com/azure/cosmos-db/table/create-table-java](https://docs.microsoft.com/azure/cosmos-db/table/create-table-java).
```
Comment (vcolin7) [inline README.md]:
```suggestion
This is the sample project for using the [azure-data-tables client library](https://search.maven.org/artifact/com.azure/azure-data-tables) with the Azure Cosmos DB Table API. The quickstart for this application is located at [https://docs.microsoft.com/azure/cosmos-db/table/create-table-java](https://docs.microsoft.com/azure/cosmos-db/table/create-table-java).
```

Comment (vcolin7) [reply README.md]: Should we point to Maven instead to our own Microsoft Docs page or GitHub repository @mario-guerra?

File: README.md [hunk 14]
```diff
@@ -1,57 +1,52 @@
...
 
 ## Features
 
-This project framework provides the following features:
+This project demonstrates how to access the Cosmos DB Table API from a java application using the Azure.Data.Tables package in the Azure SDK.  The Cosmos DB Table API is a schemaless NoSQL data store used to store tabular data.  Each row in an Azure Cosmos DB table has a partition key and row key which together form a unique key for the row.  All other properties (columns) of the row are dynamically added when objects are stored in the table.  In this way, there is no need to specify the properties (columns) of a table ahead of time as they are automatically added as needed.  If an object is loaded to the table that has new properties, the corresponding properties on the table will created automatically.
```
Comment (vcolin7) [inline README.md]:
```suggestion
This project demonstrates how to access the Cosmos DB Table API from a Java application using the `azure-data-tables` package in the Azure SDK. The Cosmos DB Table API is a schemaless NoSQL data store used to store tabular data. Each row in an Azure Cosmos DB table has a partition key and row key which together form a unique key for the row. All other properties (columns) of the row are dynamically added when objects are stored in the table. In this way, there is no need to specify the properties (columns) of a table ahead of time as they are automatically added as needed. If an object is loaded to the table that has new properties, the corresponding properties on the table will created automatically.
```

File: README.md [hunk 15]
```diff
@@ -1,57 +1,52 @@
...
 
-* Feature 1
-* Feature 2
-* ...
+This sample application is based around storing weather data.  A base set of properties (temperature, humidity, pressure, wind direction, and wind speed) are stored for each observation loaded to a table.  However, the GUI can be used to create and store weather observations with additional properties (cloud cover, UV index, air quality index, etc) to demonstrate how the Table API automatically adds properties to a table as needed.  The ability to automatically extend a table in this way is useful for any application where the data sent to the application may dynamically change at any time.
```
Comment (vcolin7) [inline README.md]:
```suggestion
This sample application is based around storing weather data. A base set of properties (temperature, humidity, pressure, wind direction, and wind speed) are stored for each observation and loaded to a table. However, the GUI can be used to create and store weather observations with additional properties (cloud cover, UV index, air quality index, etc.) to demonstrate how the Table API automatically adds properties to a table as needed. The ability to automatically extend a table in this way is useful for any application where the data sent to the application may dynamically change at any time.
```

File: README.md [hunk 16]
```diff
@@ -1,57 +1,52 @@
...
-
-To run the demo, follow these steps:
-
-(Add steps to start up the demo)
-
-1.
-2.
-3.
+This repository contains three subdirectories.
+- *starter app* - A starter version of the application without any Azure code.  Intended to be used when following the Quickstart at [https://docs.microsoft.com/azure/cosmos-db/table/create-table-java](https://docs.microsoft.com/azure/cosmos-db/table/create-table-java).
```
Comment (vcolin7) [inline README.md]:
```suggestion
- *starter app* - A starter version of the application without any Azure code. Intended to be used when following the Quickstart at [https://docs.microsoft.com/azure/cosmos-db/table/create-table-java](https://docs.microsoft.com/azure/cosmos-db/table/create-table-java).
```

File: README.md [hunk 17]
```diff
@@ -1,57 +1,52 @@
...
+- *completed app* - A completed version of the app for those who want to see the finished code or debug through a fully working application.
+- *scripts* - Azure CLI and Azure PowerShell scripts to create the Azure resources needed to run this sample.
+
+The fastest way to get started is to run the completed application.  For the application to work, you still need provision the appropriate Azure resources and use the appropriate connection string in the sample application.
```
Comment (vcolin7) [inline README.md]:
```suggestion
The fastest way to get started is to run the completed application. For the application to work, you still need provision the appropriate Azure resources and use the appropriate connection string in the sample application.
```

File: README.md [hunk 18]
```diff
@@ -1,57 +1,52 @@
...
+
+1. Clone or download the sample application from https://github.com/Azure-Samples/msdocs-azure-data-tables-sdk-java.git
+2. Use either the Azure CLI or Azure PowerShell script in the scripts directory to create the appropriate Azure resources.  If you do not have Azure CLI or Azure PowerShell installed, you may run either of these scripts in the [Azure Cloud Shell](https://shell.azure.com).
```
Comment (vcolin7) [inline README.md]:
```suggestion
2. Use either the Azure CLI or Azure PowerShell script in the scripts directory to create the appropriate Azure resources. If you do not have Azure CLI or Azure PowerShell installed, you may run either of these scripts in the [Azure Cloud Shell](https://shell.azure.com).
```

File: README.md [hunk 19]
```diff
@@ -1,57 +1,52 @@
...
+3. Each script will return the appropriate connection string for the sample application to use. Switch to folder `2-completed-app`.Then, add the value of the `connection string` in `pom.xml` file.
```
Comment (vcolin7) [inline README.md]:
```suggestion
3. Each script will return the appropriate connection string for the sample application to use. Switch to folder `2-completed-app`. Then, add the value of the `connection string` in `pom.xml` file.
```

File: README.md [hunk 20]
```diff
@@ -1,57 +1,52 @@
...
+				<azure.tables.connection.string>
+					<![CDATA[YOUR-TABLE-SERVICE-CONNECTION-STRING]]>
+				</azure.tables.connection.string>
+				<azure.tables.tableName>WeatherData</azure.tables.tableName>
+			</properties>
+			<activation>
+				<activeByDefault>true</activeByDefault>
+			</activation>
+		</profile>
+	</profiles>
```
Comment (vcolin7) [inline README.md]: Using tabs might make this render strangely in some applications.
```suggestion
<profiles>
    <profile>
        <id>local</id>
	<properties>
	    <azure.tables.connection.string>
	        <![CDATA[YOUR-TABLE-SERVICE-CONNECTION-STRING]]>
	    </azure.tables.connection.string>
            <azure.tables.tableName>WeatherData</azure.tables.tableName>
        </properties>
        <activation>
	    <activeByDefault>true</activeByDefault>
	</activation>
    </profile>
</profiles>
```

**Review by vcolin7 — COMMENTED** [review]
All the comments I added on the starter app apply to the completed app as well and vice-versa.

On another note, should we add a license header to all our files like we do in our SDKs @mario-guerra? Here's [an example](https://github.com/Azure/azure-sdk-for-java/blob/0ee602f699bdc7b39ffc3e8280f58ec59dc8f0f9/sdk/tables/azure-data-tables/src/main/java/com/azure/data/tables/TableClient.java#L1) in the Azure Data Tables library for Java.

Also, would it be better to rename this application's package to something other than `com.data.tables`? It could get easily confused with our own `com.azure.data.tables`. What about   something like `com.azure.samples.tablesapp`?

File: README.md (hunk 20 above)
Comment (v-hongli1) [reply README.md]: According to your comments, it has been fixed in the new version.

**Review by DavidCBerry13 — APPROVED** [review]

**Comment by mario-guerra** [comment]
@v-xuto please add the following copyright notice to all the source files you created for this project:

// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

**Review by vcolin7 — APPROVED** [review]
Overall it looks good to me. The only minor thing I would consider is the package name using `samples` instead of `sample` and `tablesapp` instead of `tableapp`, so it would looks this `com.azure.samples.tablesapp`. Not required but sits a bit better with me.

**Comment by v-hongli1** [comment]
The package name rename to `com.azure.samples.tablesapp`,  and the copyright notice has been added to all the source files for this project.

File: 1-starter-app/AzureTablesDemoApplication/src/main/java/com/azure/samples/tablesapp/untils/WeatherDataUtils.java [hunk 21]
```diff
@@ -0,0 +1,147 @@
+package com.azure.samples.tablesapp.untils;
```
Comment (vcolin7) [inline 1-starter-app/AzureTablesDemoApplication/src/main/java/com/azure/samples/tablesapp/untils/WeatherDataUtils.java:4]: This file is missing the copyright notice.

Comment (v-hongli1) [reply 1-starter-app/AzureTablesDemoApplication/src/main/java/com/azure/samples/tablesapp/untils/WeatherDataUtils.java:4]: Sorry, I missed fix this file. Now fixed in the new version.

**Comment by v-xuto** [comment]
@vcolin7 We have updated by your comments, please re-review this PR.
If there is no problem, please approve this PR. Thanks a lot.

**Review by vcolin7 — APPROVED** [review]

**Comment by v-xuto** [comment]
@mario-guerra Do you still have any questions about this PR? If not, can I merge currently? 
Because Vcolin7 has approved it.

**Comment by mario-guerra** [comment]
@v-xuto PR has been merged, thanks for your hard work on this sample.

**Pull request merged**